File: quant1x/exchange/security.rs
use crate::runtime::RollingOnce;
use crate::timestamp::Timestamp;
use once_cell::sync::Lazy;
use std::collections::HashMap;
use std::path::PathBuf;
use std::sync::{Arc, Mutex};
// protocol-level operations are handled inside level1::security_list::fetch_security_list

static GLOBAL_SECURITY_MAP: Lazy<Mutex<HashMap<String, SecurityInfo>>> =
    Lazy::new(|| Mutex::new(HashMap::new()));

// Global RollingOnce for securities (daily reset at 09:00)
static GLOBAL_SECURITY_ONCE: Lazy<Arc<RollingOnce>> = Lazy::new(|| {
    // create marker next to security filename (parent dir) with name security.updated
    let mut marker = PathBuf::from(crate::config::get_security_filename());
    if let Some(parent) = marker.parent() {
        marker = parent.to_path_buf();
        marker.push("security.updated");
    }
    RollingOnce::with_daily_reset(marker, 9, 0)
});

#[derive(Clone, Debug)]
pub struct SecurityInfo {
    pub code: String,
    pub name: String,
    pub lot_size: u16,
    pub price_precision: u8,
}

impl SecurityInfo {
    pub fn new(code: String, name: String, lot_size: u16, price_precision: u8) -> Self {
        Self {
            code,
            name,
            lot_size,
            price_precision,
        }
    }
}

fn init_securities_impl() -> Result<(), Box<dyn std::error::Error>> {
    // Determine cache filename
    let filename = crate::config::get_security_filename();
    let path = PathBuf::from(&filename);

    // Check last modified time and compare to today's pre-market
    let need_update = match std::fs::metadata(&path) {
        Ok(meta) => {
            if let Ok(mtime) = meta.modified() {
                let msecs = mtime
                    .duration_since(std::time::UNIX_EPOCH)
                    .map(|d| d.as_millis() as i64)
                    .unwrap_or(0);
                let cache_time = Timestamp::from(msecs);
                let check_point = Timestamp::now()
                    .pre_market_time_from_current()
                    .unwrap_or(Timestamp::now());
                // if now >= check_point and cache_time < check_point then stale
                let now = Timestamp::now();
                if now >= check_point && cache_time < check_point {
                    true
                } else {
                    false
                }
            } else {
                true
            }
        }
        Err(_) => true,
    };

    if need_update {
        log::info!("security: updating securities list into {}", filename);

        use crate::exchange::{MARKET_FLAG_SH, MARKET_FLAG_SZ, MARKET_SHANGHAI, MARKET_SHENZHEN};

        // markets to query (SZ then SH) mirror previous behavior
        let markets: [u16; 2] = [MARKET_SHENZHEN as u16, MARKET_SHANGHAI as u16];
        let mut all: Vec<(String, u16, u8, String, f64)> = Vec::new();

        for &market in markets.iter() {
            let mut start: u16 = 0;
            loop {
                match crate::level1::security_list::fetch_security_list(market as u16, start) {
                    Some(resp) => {
                        log::info!(
                            "security list market={} start={} count={}",
                            market,
                            start,
                            resp.count
                        );
                        let cnt = resp.list.len();
                        for e in resp.list.into_iter() {
                            let prefix = if market == MARKET_SHANGHAI as u16 {
                                MARKET_FLAG_SH
                            } else {
                                MARKET_FLAG_SZ
                            };
                            let code = format!("{}{}", prefix, e.code);
                            all.push((code, e.vol_unit, e.decimal_point, e.name, e.pre_close));
                        }
                        // security_list_max in C++ is 1000
                        if cnt < 1000 {
                            break;
                        }
                        start = start.wrapping_add(1000u16);
                    }
                    None => {
                        log::error!(
                            "security list request failed for market {} start {}",
                            market,
                            start
                        );
                        break;
                    }
                }
            }
        }

        // Write CSV with header (Code, VolUnit, DecimalPoint, Name, PreClose)
        fn csv_escape(s: &str) -> String {
            if s.contains(',') || s.contains('"') || s.contains('\n') {
                let mut out = String::from("\"");
                out.push_str(&s.replace("\"", "\"\""));
                out.push('"');
                out
            } else {
                s.to_string()
            }
        }

        if !all.is_empty() {
            if let Some(parent) = std::path::Path::new(&filename).parent() {
                let _ = std::fs::create_dir_all(parent);
            }
            let mut tmp = tempfile::NamedTempFile::new_in(
                std::path::Path::new(&filename)
                    .parent()
                    .unwrap_or_else(|| std::path::Path::new(".")),
            )?;
            use std::io::Write;
            writeln!(tmp.as_file_mut(), "Code,VolUnit,DecimalPoint,Name,PreClose")?;
            for s in all.iter() {
                writeln!(
                    tmp.as_file_mut(),
                    "{},{},{},{},{}",
                    s.0,
                    s.1,
                    s.2,
                    csv_escape(&s.3),
                    s.4
                )?;
            }
            let _ = tmp.persist(&filename);
        }
    }

    // Load CSV into memory and parse fields: Code,VolUnit,DecimalPoint,Name,PreClose
    if let Ok(file) = std::fs::File::open(&filename) {
        let mut rdr = csv::ReaderBuilder::new()
            .has_headers(true)
            .flexible(true)
            .trim(csv::Trim::All)
            .from_reader(file);

        let mut map = GLOBAL_SECURITY_MAP.lock().unwrap();
        map.clear();

        for result in rdr.records() {
            if let Ok(rec) = result {
                // rec: Code,VolUnit,DecimalPoint,Name,PreClose (Name may contain commas)
                let code = rec.get(0).unwrap_or("").trim().to_string();
                if code.is_empty() {
                    continue;
                }
                let vol_unit = rec.get(1).and_then(|s| s.parse::<u16>().ok()).unwrap_or(0);
                let decimal_point = rec.get(2).and_then(|s| s.parse::<u8>().ok()).unwrap_or(0);
                let name = rec.get(3).unwrap_or("").to_string();

                let code_fixed = correct_security_code(&code);
                let info = SecurityInfo::new(code_fixed.clone(), name, vol_unit, decimal_point);
                map.insert(code_fixed, info);
            }
        }
    }

    Ok(())
}

pub fn init_securities() {
    // kept for compatibility: run the init logic directly (non-once)
    let _ = init_securities_impl();
}

pub fn get_security_info(code: &str) -> Option<SecurityInfo> {
    // ensure initialized once (C++ calls global_security_once->Do(init_securities) here)
    let _ = GLOBAL_SECURITY_ONCE.do_once_try(|| -> Result<(), Box<dyn std::error::Error>> {
        init_securities_impl()?;
        Ok(())
    });

    let code_fixed = correct_security_code(code);
    let map = GLOBAL_SECURITY_MAP.lock().unwrap();
    map.get(&code_fixed).cloned()
}

/// Basic normalization similar to C++/Python CorrectSecurityCode
fn correct_security_code(s: &str) -> String {
    let v = s.trim().to_lowercase();
    // handle forms like 600519.sh or 600519.SH
    if v.contains('.') {
        let parts: Vec<&str> = v.split('.').collect();
        if parts.len() >= 2 {
            let code = parts[0];
            let suf = parts[1];
            if suf == "sh" {
                return format!("sh{:0>6}", code);
            } else if suf == "sz" {
                return format!("sz{:0>6}", code);
            }
        }
    }
    // if starts with sh/sz
    if v.starts_with("sh") || v.starts_with("sz") {
        let prefix = &v[..2];
        let rest = v[2..].to_string();
        return format!("{}{:0>6}", prefix, rest);
    }
    // plain numeric code: infer market by leading digit
    let numeric = v.chars().filter(|c| c.is_ascii_digit()).collect::<String>();
    if numeric.starts_with('6') {
        format!("sh{:0>6}", numeric)
    } else {
        format!("sz{:0>6}", numeric)
    }
}

File: quant1x/exchange/security.cpp
#include <quant1x/exchange/security.h>
#include <quant1x/std/time.h>
#include <quant1x/runtime/config.h>
#include <quant1x/exchange/blocks.h>
#include <quant1x/level1/client.h>

namespace exchange {

    // 重载<<运算符
    std::ostream& operator<<(std::ostream& os, const SecurityInfo& p) {
        os << "SecurityInfo{code: " << p.code
           << ", name: " << p.name
           << ", lotSize: " << p.lotSize
           << ", pricePrecision: " << static_cast<int>(p.pricePrecision)
           << "}";
        return os;
    }

    static inline auto global_security_once = RollingOnce::create("exchange-security", cron_expr_daily_9am);
    static inline tsl::robin_map<std::string, SecurityInfo> global_security_map = {};
    static inline std::string cache_security_filename = config::get_security_filename();

    void init_securities() {
        spdlog::debug("{}, begin", __FUNCTION__);
        auto cache_security_tp = io::last_modified_time(cache_security_filename);
        auto ec = util::check_filepath(cache_security_filename, true);
        ec.clear();
        std::string cache_security_time = exchange::timestamp(cache_security_tp).toString();
        std::string check_time_point = exchange::timestamp::now().pre_market_time().toString();
        auto now = exchange::timestamp::now().toString();
        bool bUpdate = false;
        if (cache_security_tp == 0 || cache_security_time.empty()) {
            spdlog::debug("文件[{}]不存在", cache_security_filename);
            bUpdate = true;
        } else if (now >= check_time_point && cache_security_time < check_time_point) {
            // 文件时间过了时间检查点, 且缓存文件时间比检查点还要早, 则需要同步缓存文件
            spdlog::debug("文件[{}]过时了", cache_security_filename);
            bUpdate = true;
        }
        //bUpdate = true;
        if (bUpdate) {
            // 同步证券信息文件
            try {
                // 开始业务处理
                auto markets = {exchange::MarketType::ShangHai, exchange::MarketType::ShenZhen};
                // 1. 获取证券列表
                std::vector<level1::Security> allSecurity;
                for(auto const & marketId : markets) {
                    std::string prefix = exchange::GetMarketFlag(marketId);
                    int start = 0;
                    for(;;){
                        level1::SecurityListRequest reqSecurityList((int) marketId, start);
                        level1::SecurityListResponse respSecurityList;
                        auto conn = level1::client();
                        auto err = level1::process(conn->socket(), reqSecurityList, respSecurityList);
                        if (err) {
                            throw std::runtime_error(fmt::format("Process error: {}", err.message()));
                        }
                        if (!respSecurityList.List.empty()) {
                            //allSecurity.insert(allSecurity.end(), respSecurityList.List.begin(), respSecurityList.List.end());
                            for(int i = 0; i < respSecurityList.Count; ++i) {
                                auto v = &respSecurityList.List[i];
                                v->Code = prefix + v->Code;
                                if(!exchange::checkIndexAndStock(v->Code)) {
                                    continue;
                                }
                                if(exchange::AssertBlockBySecurityCode(&(v->Code))) {
                                    auto blk  = exchange::get_sector_info(v->Code);
                                    if(blk.has_value()) {
                                        v->Name = (*blk).name;
                                    }
                                }
                                allSecurity.emplace_back(*v);
                            }
                        }
                        if (respSecurityList.List.size() < level1::security_list_max) {
                            break;
                        }
                        start += level1::security_list_max;
                    }
                }
                if (!allSecurity.empty()){
                    io::CSVWriter writer(cache_security_filename);
                    writer.write_row("Code", "VolUnit", "DecimalPoint", "Name", "PreClose");
                    for (auto const &v: allSecurity) {
                        writer.write_row(v.Code, v.VolUnit, v.DecimalPoint, v.Name, v.PreClose);
                    }
                }
            }
            catch (const std::exception& e) {
                spdlog::error("Error: {}", e.what());
            }

        }
        {
            // 其它情况加载缓存
            io::CSVReader<4> in(cache_security_filename);
            in.read_header(io::ignore_extra_column, "Code", "VolUnit", "DecimalPoint", "Name");
            std::string code;
            uint16_t lotSize = 0;
            uint8_t pricePrecision = 0;
            std::string name;
            global_security_map.clear();
            while (in.read_row(code, lotSize, pricePrecision, name)) {
                auto v = SecurityInfo{code, name, lotSize, pricePrecision};
                global_security_map.insert_or_assign(code, v);
            }
        }
        spdlog::debug("{}, end", __FUNCTION__);
    }

    std::optional<SecurityInfo> get_security_info(const std::string &code) {
        global_security_once->Do(init_securities);
        auto securityCode = exchange::CorrectSecurityCode(code);
        auto it = global_security_map.find(securityCode);
        if (it != global_security_map.end()) {
            return it->second; // 返回指针
        } else {
            return std::nullopt; // 未找到返回空指针
        }
    }

    namespace {
        constexpr double HighLimit = 0.20;
        constexpr double NormalLimit = 0.10;
        constexpr double BeijingLimit = 0.30;
    }

    // 工具函数：检查字符串是否以指定前缀开头
    bool starts_with(std::string_view str, std::string_view prefix) {
        return str.size() >= prefix.size() &&
               str.substr(0, prefix.size()) == prefix;
    }

    /**
     * @brief 获取证券的涨停板比率（例如：0.10 表示 10%）
     * @param security_code 证券代码（如 "600000"）
     * @return 涨停幅度（如主板返回 0.10，创业板返回 0.20）
     */
    double get_up_limit_rate(const std::string& security_code) {
        auto [mid, mcode, symbol] = DetectMarket(security_code);

        if (mcode == market_beijing) {
            return BeijingLimit; // 北交所特殊处理
        }

        std::string_view symbol_view = symbol;

        // 判断是否是创业板/科创板
        static constexpr std::array<std::string_view, 2> kHighLimitPrefixes = {"30", "68"};
        for (const auto& prefix : kHighLimitPrefixes) {
            if (starts_with(symbol_view, prefix)) {
                return HighLimit; // 高涨跌幅限制
            }
        }

        return NormalLimit; // 默认涨幅限制
    }

    /**
     * @brief 根据昨日收盘价和证券代码，计算涨停价格
     * @param security_code
     * @param prev_close
     * @return
     */
    double calc_limit_up_price(const std::string& security_code, double prev_close) {
        double rate = get_up_limit_rate(security_code);
        auto price = numerics::decimal(prev_close);
        auto result = price * (1.0 + rate);
        return numerics::decimal(result);
    }

} // namespace exchange